File: model-client/src/main/java/org/modelix/model/client/RestWebModelClient.java
package org.modelix.model.client;

import org.modelix.model.IKeyListener;
import org.modelix.model.IKeyValueStore;
import org.modelix.model.api.IIdGenerator;
import org.apache.commons.io.FileUtils;
import org.apache.log4j.Level;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.json.JSONArray;
import org.json.JSONObject;
import org.modelix.model.util.StreamUtils;
import org.modelix.model.KeyValueStoreCache;
import org.modelix.model.lazy.IDeserializingKeyValueStore;
import org.modelix.model.lazy.ObjectStoreCache;
import org.modelix.model.persistent.HashUtil;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.ClientRequestContext;
import javax.ws.rs.client.ClientRequestFilter;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.sse.InboundSseEvent;
import javax.ws.rs.sse.SseEventSource;
import java.io.File;
import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.stream.Stream;

public class RestWebModelClient implements IModelClient {
  private static final Logger LOG = LogManager.getLogger(RestWebModelClient.class);
  public static final String MODEL_URI_VAR_NAME = "MODEL_URI";
  private static String defaultToken;

  static {
    try {
      File sharedSecretFile = new File("/secrets/modelsecret/modelsecret.txt");
      if (sharedSecretFile.exists()) {
        defaultToken = FileUtils.readFileToString(sharedSecretFile, StandardCharsets.UTF_8);
      }
    } catch (Exception ex) {
      if (LOG.isEnabledFor(Level.ERROR)) {
        LOG.error("Failed to load default token", ex);
      }
    }
  }

  public static String getModelUrlFromEnv() {
    String url = System.getProperty(MODEL_URI_VAR_NAME);
    if ((url == null || url.length() == 0)) {
      url = System.getenv(MODEL_URI_VAR_NAME);
    }
    return url;
  }

  public static String getDefaultUrl() {
    String urlFromEnv = getModelUrlFromEnv();
    if ((urlFromEnv != null && urlFromEnv.length() > 0)) {
      return urlFromEnv;
    } else {
      return "http://modelix.q60.de:80/model/";
    }
  }

  private int clientId = 0;
  private Client client;
  private final List<SseListener> listeners = new ArrayList<>();
  private String baseUrl;
  private IKeyValueStore asyncStore = new GarbageFilteringStore(new AsyncStore(this));
  private ObjectStoreCache cache = new ObjectStoreCache(new KeyValueStoreCache(asyncStore));
  private IdGenerator idGenerator;
  private ScheduledFuture<?> watchDogTask;
  private String authToken = defaultToken;

  public RestWebModelClient() {
    this(null);
  }

  public RestWebModelClient(String baseUrl) {
    this.baseUrl = baseUrl;
    if (isEmptyString(this.baseUrl)) {
      this.baseUrl = getDefaultUrl();
    }
    if (!(this.baseUrl.endsWith("/"))) {
      this.baseUrl += "/";
    }

    client = ClientBuilder.newBuilder().register(new ClientRequestFilter() {
      @Override
      public void filter(ClientRequestContext ctx) throws IOException {
        ctx.getHeaders().add(HttpHeaders.AUTHORIZATION, "Bearer " + authToken);
      }
    }).build();
    watchDogTask = SharedExecutors.fixDelay(1000, new Runnable() {
      public void run() {
        List<SseListener> ls;
        synchronized (listeners) {
          ls = new ArrayList<SseListener>(listeners);
        }
        for (SseListener l : ls) {
          try {
            l.ensureConnected();
          } catch (Exception ex) {
            if (LOG.isEnabledFor(Level.ERROR)) {
              LOG.error("", ex);
            }
          }
        }
      }
    });
  }

  public String getBaseUrl() {
    return baseUrl;
  }

  public void dispose() {
    watchDogTask.cancel(false);
    synchronized (listeners) {
      listeners.forEach(SseListener::dispose);
      listeners.clear();
    }
  }

  @Override
  public String get(String key) {
    boolean isHash = HashUtil.isSha256(key);
    if (isHash) {
      if (LOG.isDebugEnabled()) {
        LOG.debug("GET " + key);
      }
    }
    long start = System.currentTimeMillis();
    Response response = client.target(baseUrl + "get/" + URLEncoder.encode(key, StandardCharsets.UTF_8)).request().buildGet().invoke();
    if (response.getStatus() == Response.Status.OK.getStatusCode()) {
      String value = response.readEntity(String.class);
      long end = System.currentTimeMillis();
      if (isHash) {
        if (LOG.isDebugEnabled()) {
          LOG.debug("GET " + key + " took " + (end - start) + " ms: " + value);
        }
      }
      return value;
    } else if (response.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
      return null;
    } else {
      throw new RuntimeException("Request for key '" + key + "' failed: " + response.getStatusInfo());
    }
  }

  @Override
  public Map<String, String> getAll(Iterable<String> keys) {
    if (!keys.iterator().hasNext()) {
      return new HashMap<>();
    }

    JSONArray json = new JSONArray();
    for (String key : keys) {
      json.put(key);
    }
    String body = json.toString();

    long start = System.currentTimeMillis();
    Response response = client.target(baseUrl + "getAll").request(MediaType.APPLICATION_JSON).put(Entity.text(body));
    if (response.getStatus() == Response.Status.OK.getStatusCode()) {
      String jsonStr = response.readEntity(String.class);
      JSONArray responseJson = new JSONArray(jsonStr);
      Map<String, String> result = new LinkedHashMap<>(16, (float) 0.75, false);
      for (Object entry_ : responseJson) {
        JSONObject entry = (JSONObject) entry_;
        result.put(entry.getString("key"), entry.optString("value"));
      }
      long end = System.currentTimeMillis();
      return result;
    } else {
      throw new RuntimeException(String.format("Request for %d keys failed (%s, ...): %s",
              keys.spliterator().getExactSizeIfKnown(),
              StreamUtils.toStream(keys).findFirst().orElse(null),
              response.getStatusInfo()));
    }
  }

  public void setAuthToken(String token) {
    this.authToken = token;
  }

  public String getEmail() {
    Response response = client.target(baseUrl + "getEmail").request().buildGet().invoke();
    if (response.getStatus() == Response.Status.OK.getStatusCode()) {
      return response.readEntity(String.class);
    } else {
      throw new RuntimeException("Request for e-mail address failed: " + response.getStatusInfo());
    }
  }

  @Override
  public void listen(final String key, final IKeyListener keyListener) {
    SseListener sseListener = new SseListener(key, keyListener);
    synchronized (listeners) {
      listeners.add(sseListener);
    }
  }

  @Override
  public void removeListener(final String key, final IKeyListener listener) {
    synchronized (listeners) {
      listeners.removeIf(it -> Objects.equals(it.key, key) && it.keyListener == listener);
    }
  }

  @Override
  public void put(String key, String value) {
    if (!(key.matches("[a-zA-Z0-9-_]{43}"))) {
      if (LOG.isDebugEnabled()) {
        LOG.debug("PUT " + key + " = " + value);
      }
    }
    Response response = client.target(baseUrl + "put/" + URLEncoder.encode(key, StandardCharsets.UTF_8)).request(MediaType.TEXT_PLAIN).put(Entity.text(value));
    if (response.getStatusInfo().getFamily() != Response.Status.Family.SUCCESSFUL) {
      throw new RuntimeException("Failed to store entry (" + response.getStatusInfo() + ") " + key + " = " + value);
    }
  }

  @Override
  public void putAll(final Map<String, String> entries) {
    Consumer<JSONArray> sendBatch = new Consumer<JSONArray>() {
      public void accept(JSONArray json) {
        if (LOG.isDebugEnabled()) {
          LOG.debug("PUT batch of " + json.length() + " entries");
        }
        Response response = client.target(baseUrl + "putAll").request(MediaType.APPLICATION_JSON).put(Entity.text(json.toString()));
        if (response.getStatusInfo().getFamily() != Response.Status.Family.SUCCESSFUL) {
          throw new RuntimeException(String.format("Failed to store %d entries (%s) %s",
                  entries.size(),
                  response.getStatusInfo(),
                  entries.entrySet().stream().map(e -> e.getKey() + " = " + e.getValue() + ", ...").findFirst().orElse("")));
        }
      }
    };

    if (LOG.isDebugEnabled()) {
      LOG.debug("PUT " + entries.size() + " entries");
    }
    JSONArray json = new JSONArray();
    int approxSize = 0;
    for (var entry : entries.entrySet()) {
      JSONObject jsonEntry = new JSONObject();
      jsonEntry.put("key", entry.getKey());
      jsonEntry.put("value", entry.getValue());
      approxSize += entry.getKey().length();
      approxSize += entry.getValue().length();
      json.put(jsonEntry);
      if (!(entry.getKey().matches("[a-zA-Z0-9-_]{43}"))) {
        if (LOG.isDebugEnabled()) {
          LOG.debug("PUT " + entry.getKey() + " = " + entry.getValue());
        }
      }

      if (json.length() >= 5000 || approxSize > 10000000) {
        sendBatch.accept(json);
        json = new JSONArray();
        approxSize = 0;
      }
    }
    if (json.length() > 0) {
      sendBatch.accept(json);
    }
  }

  @Override
  public void prefetch(String key) {
  }

  public int getClientId() {
    if (clientId == 0) {
      Response response = client.target(baseUrl + "counter/clientId").request().post(Entity.text(""));
      String idStr = response.readEntity(String.class);
      clientId = Integer.parseInt(idStr);
    }
    return clientId;
  }

  @Override
  public IDeserializingKeyValueStore getStoreCache() {
    return cache;
  }

  @Override
  public IKeyValueStore getAsyncStore() {
    return asyncStore;
  }

  @Override
  public synchronized IIdGenerator getIdGenerator() {
    if (idGenerator == null) {
      idGenerator = new IdGenerator(getClientId());
    }
    return idGenerator;
  }

  public class SseListener {
    private String key;
    private IKeyListener keyListener;
    private final Object notificationLock = new Object();
    private String lastValue;
    private Sse[] sse = new Sse[2];
    private boolean disposed = false;

    public SseListener(final String key, IKeyListener listener) {
      this.key = key;
      this.keyListener = listener;
    }
    public void dispose() {
      if (disposed) {
        return;
      }
      disposed = true;
      for (int i = 0; i < sse.length; i++) {
        if (sse[i] != null) {
          sse[i].sse.close();
          sse[i] = null;
        }
      }
    }
    public synchronized void ensureConnected() {
      if (disposed) {
        return;
      }
      for (int i = 0; i < sse.length; i++) {
        if (sse[i] == null) {
          continue;
        }
        if (sse[i].birth > System.currentTimeMillis()) {
          sse[i].birth = System.currentTimeMillis();
        }
        if (!(sse[i].sse.isOpen())) {
          sse[i] = null;
        }
      }

      for (int i = 0; i < sse.length; i++) {
        // To support rebalancing after scaling the cluster a connection shouldn't be open for too long. 
        if (sse[i] != null && sse[i].getAge() > 20000) {
          sse[i].sse.close();
          sse[i] = null;
        }
      }

      long youngest = Stream.of(sse).filter(Objects::nonNull).mapToLong(it -> it.birth).reduce(0L, Math::max);

      if (System.currentTimeMillis() - youngest < 5000) {
        return;
      }

      for (int i = 0; i < sse.length; i++) {
        if (sse[i] != null) {
          continue;
        }
        String url = baseUrl + "subscribe/" + URLEncoder.encode(key, StandardCharsets.UTF_8);
        if (LOG.isTraceEnabled()) {
          LOG.trace("Connecting to " + url);
        }
        WebTarget target = client.target(url);
        sse[i] = new Sse(SseEventSource.target(target).reconnectingEvery(1, TimeUnit.SECONDS).build());
        sse[i].sse.register(new Consumer<InboundSseEvent>() {
          @Override
          public void accept(InboundSseEvent event) {
            String value = event.readData();
            synchronized (notificationLock) {
              if (!(Objects.equals(value, lastValue))) {
                lastValue = value;
                keyListener.changed(key, value);
              }
            }
          }
        }, new Consumer<Throwable>() {
          @Override
          public void accept(Throwable ex) {
            if (LOG.isEnabledFor(Level.ERROR)) {
              LOG.error("", ex);
            }
          }
        });
        if (disposed) {
          return;
        }
        sse[i].sse.open();
        if (LOG.isTraceEnabled()) {
          LOG.trace("Connected to " + url);
        }
        break;
      }
    }
    public class Sse {
      private long birth = System.currentTimeMillis();
      private SseEventSource sse;
      public Sse(SseEventSource sse) {
        this.sse = sse;
      }
      public long getAge() {
        return System.currentTimeMillis() - birth;
      }
    }
  }

  private static boolean isEmptyString(String str) {
    return str == null || str.length() == 0;
  }
}

File: model-client/src/jvmMain/kotlin/org/modelix/model/client/RestWebModelClient.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */

package org.modelix.model.client

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.auth.*
import io.ktor.client.plugins.auth.providers.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.serialization.*
import io.ktor.util.reflect.*
import io.ktor.utils.io.*
import io.ktor.utils.io.charsets.*
import io.ktor.utils.io.core.*
import kotlinx.coroutines.*
import kotlinx.coroutines.CancellationException
import org.apache.log4j.LogManager
import org.json.JSONArray
import org.json.JSONObject
import org.modelix.model.IKeyListener
import org.modelix.model.IKeyValueStore
import org.modelix.model.KeyValueStoreCache
import org.modelix.model.api.IIdGenerator
import org.modelix.model.lazy.IDeserializingKeyValueStore
import org.modelix.model.lazy.ObjectStoreCache
import org.modelix.model.lazy.PrefetchCache
import org.modelix.model.oauth.ModelixOAuthClient
import org.modelix.model.persistent.HashUtil
import org.modelix.model.sleep
import org.modelix.model.util.StreamUtils.toStream
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

val HttpResponse.successful: Boolean
    get() = this.status.value in 200..299

val HttpResponse.unsuccessful: Boolean
    get() = !successful

val HttpResponse.forbidden: Boolean
    get() = status == HttpStatusCode.Forbidden

interface ConnectionListener {
    fun receivedForbiddenResponse()
    fun receivedSuccessfulResponse()
}

typealias ConnectionStatusListener = (oldValue: RestWebModelClient.ConnectionStatus, newValue: RestWebModelClient.ConnectionStatus) -> Unit

/**
 * We need to specify the connection listeners right into the constructor because connection is started in the constructor.
 */
class RestWebModelClient @JvmOverloads constructor(
    var baseUrl: String = defaultUrl,
    val authTokenProvider: (() -> String?)? = null,
    initialConnectionListeners: List<ConnectionListener> = emptyList()
) : IModelClient {

    companion object {
        private val LOG = LogManager.getLogger(RestWebModelClient::class.java)
        const val MODEL_URI_VAR_NAME = "MODEL_URI"
        val modelUrlFromEnv: String?
            get() {
                var url = System.getProperty(MODEL_URI_VAR_NAME)
                if (url.isNullOrEmpty()) {
                    url = System.getenv(MODEL_URI_VAR_NAME)
                }
                return url
            }

        val defaultUrl: String
            get() {
                val urlFromEnv = modelUrlFromEnv
                return if (urlFromEnv.isNullOrEmpty()) {
                    "http://localhost:28101/"
                } else {
                    urlFromEnv
                }
            }
    }

    @Deprecated("Replaced by connectionStatusListeners")
    private val connectionListeners = LinkedList(initialConnectionListeners)

    override val clientId: Int
        get() {
            if (clientIdInternal == 0) {
                throw IllegalStateException("Client ID is not initialized yet. Client state: $connectionStatus")
            }
            return clientIdInternal
        }
    private var clientIdInternal: Int = 0
    private val coroutineScope = CoroutineScope(Dispatchers.Default)
    private val client = HttpClient(CIO) {
        this.followRedirects = false
        install(HttpTimeout) {
            connectTimeoutMillis = 1.seconds.inWholeMilliseconds
            requestTimeoutMillis = 30.seconds.inWholeMilliseconds
        }
        install(ContentNegotiation) {
            this.register(
                ContentType.Application.Json,
                object : ContentConverter {
                    override suspend fun deserialize(charset: Charset, typeInfo: TypeInfo, content: ByteReadChannel): Any {
                        val str = content.readRemaining().readText(charset)
                        return if (typeInfo.type == JSONArray::class) JSONArray(str) else JSONObject(str)
                    }

                    override suspend fun serialize(
                        contentType: ContentType,
                        charset: Charset,
                        typeInfo: TypeInfo,
                        value: Any
                    ): OutgoingContent {
                        return TextContent(value.toString(), contentType)
                    }
                }
            )
        }
        install(Auth) {
            bearer {
                loadTokens {
                    val tp = authTokenProvider
                    if (tp == null) {
                        ModelixOAuthClient.getTokens()?.let { BearerTokens(it.accessToken, it.refreshToken) }
                    } else {
                        val token = tp()
                        if (token == null) {
                            connectionStatus = ConnectionStatus.WAITING_FOR_TOKEN
                            null
                        } else {
                            BearerTokens(token, "")
                        }
                    }
                }
                refreshTokens {
                    val tp = authTokenProvider
                    if (tp == null) {
                        var url = baseUrl
                        if (!url.endsWith("/")) url += "/"
                        if (url.endsWith("/model/")) url = url.substringBeforeLast("/model/")
                        connectionStatus = ConnectionStatus.WAITING_FOR_TOKEN
                        val tokens = ModelixOAuthClient.authorize(url)
                        BearerTokens(tokens.accessToken, tokens.refreshToken)
                    } else {
                        val providedToken = tp()
                        if (providedToken != null && providedToken != this.oldTokens?.accessToken) {
                            BearerTokens(providedToken, "")
                        } else {
                            connectionStatus = ConnectionStatus.WAITING_FOR_TOKEN
                            null
                        }
                    }
                }
            }
        }
        install(HttpRequestRetry) {
            retryOnExceptionOrServerErrors(maxRetries = 3)
            exponentialDelay()
            modifyRequest {
                try {
//                    connectionStatus = ConnectionStatus.SERVER_ERROR
                    runBlocking {
                        response?.let { println(it.bodyAsText()) }
                    }
                } catch (e: Exception) {
                    LOG.error("", e)
                }
            }
        }
    }.apply {
        plugin(HttpSend).intercept { request ->
            val call = execute(request)
            val response = call.response
            if (response.successful) {
                receivedSuccessfulResponse()
            } else {
                if (response.status == HttpStatusCode.Unauthorized) {
                    receivedForbiddenResponse()
                }
            }
            call
        }
    }
    private val listeners: MutableList<PollingListener> = ArrayList()
    override val asyncStore: IKeyValueStore = AsyncStore(this)
    private val cache = PrefetchCache.contextIndirectCache(ObjectStoreCache(KeyValueStoreCache(asyncStore)))
    private val pendingWrites = AtomicInteger(0)
    var connectionStatus: ConnectionStatus = ConnectionStatus.NEW
        private set(value) {
            val oldValue = field
            field = value
            if (oldValue == value) return

            for (listener in connectionStatusListeners) {
                try {
                    listener(oldValue, value)
                } catch (e: Exception) {
                    LOG.error("Exception in status listener", e)
                }
            }
        }
    private var connectionStatusListeners: Set<ConnectionStatusListener> = emptySet()

    private fun startConnectionWatchdog() {
        coroutineScope.launch {
            while (isActive) {
                try {
                    connectNow()
                    if (connectionStatus == ConnectionStatus.CONNECTED) {
                        delay(10.seconds)
                    } else {
                        delay(3.seconds)
                    }
                } catch (e: CancellationException) {
                    break
                } catch (e: Exception) {
                    LOG.debug("", e)
                    if (connectionStatus == ConnectionStatus.CONNECTED) {
                        connectionStatus = ConnectionStatus.DISCONNECTED
                    }
                    delay(3.seconds)
                }
            }
        }
    }

    private suspend fun connectNow() {
        val response = client.get(baseUrl + "getEmail")
        if (response.successful || response.status == HttpStatusCode.NotFound) {
            if (clientIdInternal == 0) {
                loadClientId()
            }
            if (clientIdInternal != 0) {
                idGeneratorInternal = IdGenerator(clientIdInternal)
            }
            connectionStatus = ConnectionStatus.CONNECTED
        } else if (response.status == HttpStatusCode.Unauthorized) {
            connectionStatus = ConnectionStatus.WAITING_FOR_TOKEN
        }
    }

    fun addStatusListener(listener: ConnectionStatusListener) {
        connectionStatusListeners += listener
    }
    fun removeStatusListener(listener: ConnectionStatusListener) {
        connectionStatusListeners -= listener
    }

    private suspend fun loadClientId() {
        val targetUri = baseUrl + "counter/clientId"
        try {
            val response = client.post(targetUri)
            val body = response.bodyAsText()
            if (response.unsuccessful) {
                throw RuntimeException("Unable to get the clientId by querying $targetUri: ${response.status}\n$body")
            }
            clientIdInternal = body.toInt()
            LOG.info("Client ID received: $clientIdInternal")
        } catch (e: Exception) {
            throw RuntimeException("Unable to get the clientId by querying $targetUri", e)
        }
    }

    override val idGenerator: IIdGenerator = object : IIdGenerator {
        override fun generate(): Long {
            if (idGeneratorInternal == null) {
                runBlocking {
                    connectNow()
                }
            }
            val gen = idGeneratorInternal ?: throw IllegalStateException("Not connected yet")
            return gen.generate()
        }
    }

    private var idGeneratorInternal: IIdGenerator? = null

    override fun toString() = "RestWebModelClient($baseUrl)"

    private fun receivedForbiddenResponse() {
        connectionListeners.forEach { it.receivedForbiddenResponse() }
    }

    private fun receivedSuccessfulResponse() {
        connectionListeners.forEach { it.receivedSuccessfulResponse() }
    }

    fun addForbiddenListener(listener: ConnectionListener) {
        connectionListeners.add(listener)
    }

    fun dispose() {
        synchronized(listeners) {
            listeners.forEach { it.dispose() }
            listeners.clear()
        }
        coroutineScope.cancel("model client disposed")
    }

    override fun getPendingSize(): Int = pendingWrites.get()

    override fun get(key: String): String? {
        val isHash = HashUtil.isSha256(key)
        if (isHash) {
            if (LOG.isDebugEnabled) {
                LOG.debug("GET $key")
            }
        }
        return runBlocking {
            val start = System.currentTimeMillis()
            val uri = baseUrl + "get/" + URLEncoder.encode(key, StandardCharsets.UTF_8)
            try {
                val response = client.get(uri)
                return@runBlocking when (response.status) {
                    HttpStatusCode.OK -> {
                        val value = response.bodyAsText()
                        val end = System.currentTimeMillis()
                        if (isHash) {
                            if (LOG.isDebugEnabled) {
                                LOG.debug("GET " + key + " took " + (end - start) + " ms: " + value)
                            }
                        }
                        value
                    }
                    HttpStatusCode.NotFound -> {
                        null
                    }
                    else -> {
                        throw RuntimeException("Request for key '" + key + "' failed: " + response.status)
                    }
                }
            } catch (e: Exception) {
                throw RuntimeException("Unable to connect to '$uri' to get key $key", e)
            }
        }
    }

    override fun getAll(keys: Iterable<String>): Map<String, String?> {
        if (!keys.iterator().hasNext()) {
            return HashMap()
        }

        return runBlocking {
            val result: MutableMap<String, String?> = LinkedHashMap(16, 0.75.toFloat(), false)
            var json = JSONArray()
            val batch = suspend {
                val response = client.put(baseUrl + "getAll") {
                    setBody(json)
                    contentType(ContentType.Application.Json)
                }
                if (response.status == HttpStatusCode.OK) {
                    val jsonStr = response.bodyAsText()
                    val responseJson = JSONArray(jsonStr)
                    for (entry_: Any in responseJson) {
                        val entry = entry_ as JSONObject
                        result[entry.getString("key")] = entry.optString("value", null)
                    }
                    json = JSONArray()
                } else {
                    throw RuntimeException(
                        String.format(
                            "Request for %d keys failed (%s, ...): %s",
                            keys.spliterator().exactSizeIfKnown,
                            toStream(keys).findFirst().orElse(null),
                            response.status
                        )
                    )
                }
            }

            for (key in keys) {
                json.put(key)
                if (json.length() >= 5000) batch()
            }

            if (json.length() > 0) batch()

            return@runBlocking result
        }
    }

    fun getEmail(): String {
        return runBlocking {
            val response = client.get(baseUrl + "getEmail")
            if (response.successful) {
                response.bodyAsText()
            } else {
                throw RuntimeException("Request for e-mail address failed: " + response.status)
            }
        }
    }

    override fun listen(key: String, listener: IKeyListener) {
        val pollingListener = PollingListener(key, listener)
        synchronized(listeners) {
            listeners.add(pollingListener)
            pollingListener.start()
        }
    }

    override fun removeListener(key: String, listener: IKeyListener) {
        synchronized(listeners) {
            val toRemove = listeners.filter { it.key == key && it.keyListener === listener }
            listeners.removeAll(toRemove)
            toRemove.forEach { it.dispose() }
        }
    }

    override fun put(key: String, value: String?) {
        if (!key.matches(HashUtil.HASH_PATTERN)) {
            if (LOG.isDebugEnabled) {
                LOG.debug("PUT $key = $value")
            }
        }
        runBlocking {
            val url = baseUrl + "put/" + URLEncoder.encode(key, StandardCharsets.UTF_8)
            try {
                val response = client.put(url) {
                    setBody(value)
                }
                if (response.unsuccessful) {
                    throw RuntimeException("Failed to store entry (${response.status} ${response.status}) $key = $value. " + response.bodyAsText())
                }
            } catch (e: Exception) {
                throw RuntimeException("Failed executing a put to $url", e)
            }
        }
    }

    fun sortEntriesByDependency(unsorted: Map<String, String?>): Map<String, String?> {
        val sorted = LinkedHashMap<String, String?>()
        object {
            fun putEntry(key: String, value: String?) {
                if (sorted.containsKey(key)) return
                for (depKey in HashUtil.extractSha256(value)) {
                    if (sorted.containsKey(depKey)) continue
                    if (unsorted.containsKey(depKey)) {
                        val depValue = unsorted[depKey]
                        putEntry(depKey, depValue)
                    }
                }
                sorted[key] = value
            }

            fun putAll() {
                for (entry in unsorted) {
                    putEntry(entry.key, entry.value)
                }
            }
        }.putAll()

        return sorted
    }

    override fun putAll(entries_: Map<String, String?>) {
        val entries = sortEntriesByDependency(entries_)
        val sendBatch: suspend (JSONArray, Int) -> Unit = sendBatch@{ json: JSONArray, remaining: Int ->
            for (attempt in 1..3) {
                if (LOG.isDebugEnabled) {
                    LOG.debug("PUT batch of ${json.length()} entries, $remaining remaining")
                }
                val response = client.put(baseUrl + "putAll") {
                    setBody(json)
                    contentType(ContentType.Application.Json)
                }
                if (response.successful) {
                    return@sendBatch
                }
                val message = String.format(
                    "Failed to store %d entries (%s) %s: %s (attempt %d)",
                    entries.size,
                    response.status,
                    entries.entries.stream().map { e: Map.Entry<String?, String?> -> e.key.toString() + " = " + e.value + ", ..." }.findFirst().orElse(""),
                    response.bodyAsText(),
                    attempt
                )
                if (attempt == 3) throw RuntimeException(message) else LOG.warn(message)
                sleep(1000)
            }
        }
        if (LOG.isDebugEnabled) {
            LOG.debug("PUT " + entries.size + " entries")
        }

        runBlocking {
            var remainingEntries = entries.size
            try {
                pendingWrites.addAndGet(remainingEntries)
                var json = JSONArray()
                var approxSize = 0
                for ((key, value) in entries) {
                    val jsonEntry = JSONObject()
                    jsonEntry.put("key", key)
                    jsonEntry.put("value", value)
                    approxSize += key.length
                    approxSize += value!!.length
                    json.put(jsonEntry)
                    if (!key.matches(HashUtil.HASH_PATTERN)) {
                        if (LOG.isDebugEnabled) {
                            LOG.debug("PUT $key = $value")
                        }
                    }
                    if (json.length() >= 5000 || approxSize > 10000000) {
                        sendBatch(json, remainingEntries)
                        remainingEntries -= json.length()
                        pendingWrites.addAndGet(-json.length())
                        json = JSONArray()
                        approxSize = 0
                    }
                }
                if (json.length() > 0) {
                    sendBatch(json, remainingEntries)
                }
            } finally {
                pendingWrites.addAndGet(-remainingEntries)
            }
        }
    }

    override fun prefetch(key: String) {}

    override val storeCache: IDeserializingKeyValueStore
        get() = cache

    inner class PollingListener(val key: String, val keyListener: IKeyListener) {
        private var lastValue: String? = null
        private var job: Job? = null
        var nextDelay: Duration = 1.seconds
        fun dispose() {
            job?.cancel("listener disposed")
        }
        fun start() {
            job = coroutineScope.launch {
                while (isActive) {
                    try {
                        if (connectionStatus == ConnectionStatus.CONNECTED) {
                            run()
                        } else {
                            delay(1.seconds)
                        }
                        nextDelay = 1.seconds
                    } catch (e: CancellationException) {
                        break
                    } catch (e: Exception) {
                        LOG.error("Polling for '$key' failed", e)
                        delay(nextDelay)
                        nextDelay = (nextDelay * 1.2).coerceAtMost(2.minutes)
                    }
                }
            }
        }
        private suspend fun run() {
            var url = baseUrl + "poll/" + URLEncoder.encode(key, StandardCharsets.UTF_8)
            if (lastValue != null) {
                url += "?lastKnownValue=" + URLEncoder.encode(lastValue, StandardCharsets.UTF_8)
            }

            val value: String?
            val response = client.get(url) {
                timeout {
                    requestTimeoutMillis = 60.seconds.inWholeMilliseconds // long polling
                }
            }
            if (response.status == HttpStatusCode.NotFound) {
                delay(1.seconds)
            } else {
                if (response.unsuccessful) {
                    throw RuntimeException("Request for key '" + key + "' failed: " + response.status)
                }
                value = response.bodyAsText()
                if (value != lastValue) {
                    lastValue = value
                    keyListener.changed(key, value)
                }
            }
        }
    }

    init {
        if (baseUrl.isEmpty()) {
            baseUrl = defaultUrl
        }
        if (!(baseUrl.endsWith("/"))) {
            baseUrl += "/"
        }
        startConnectionWatchdog()
    }

    enum class ConnectionStatus {
        NEW,
        WAITING_FOR_TOKEN,
        DISCONNECTED,
//        SERVER_ERROR,
        CONNECTED;
    }
}
